# Exploring Used Car Listings using eBay Kleinanzeign Data




Name: Bryan B.  
Date: 3/9/2022
Updated: 6/11/2022


## Contents
[Introduction](#1.)  
[Load the Dataset](#2.)  
[Clean the Columns](#3.)  
[Initial Data Exploration and Cleaning](#4.)  
[Further Explore the Odometer and Price Columns](#5.)  
[Exploring the Date Columns](#6.)  
[Dealing with Incorrect Registration Year Data](#7.)  
[Exploring Price by Brand](#8.)  
[Conclusion](#9.)

## <a name="1."></a> 1. Introduction

Project to explore and understand used car listing data from [eBay Kleinanzeigen](https://www.ebay-kleinanzeigen.de/), a classifieds section of the German eBay website.

The goal of this project is to clean and analyze the car listings data using Python.

## <a name="2."></a> 2. Load the Dataset 

Source: https://data.world/data-society/used-cars-data  
The dataset was originally scraped and uploaded to Kaggle.  The original dataset is not available on Kaggle currently, but can be found [here](https://data.world/data-society/used-cars-data).  
The version of the dataset used in this project is a sample of 50,000 data points prepared by Dataquest including simulating a less-cleaned version of the data.

The Data Dictionary provided with the dataset is provided below:  

Data Dictionary:  
- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.  
- `name` - Name of the car.  
- `seller` - Whether the seller is private or a dealer.  
- `offerType` - The type of listing  
- `price` - The price on the ad to sell the car.  
- `abtest` - Whether the listing is included in an A/B test.  
- `vehicleType` - The vehicle Type.  
- `yearOfRegistration` - The year in which which year the car was first registered.  
- `gearbox` - The transmission type.  
- `powerPS` - The power of the car in PS.  
- `model` - The car model name.  
- `kilometer` - How many kilometers the car has driven.  
- `monthOfRegistration` - The month in which which year the car was first registered.  
- `fuelType` - What type of fuel the car uses.  
- `brand` - The brand of the car.  
- `notRepairedDamage` - If the car has a damage which is not yet repaired.  
- `dateCreated` - The date on which the eBay listing was created.  
- `nrOfPictures` - The number of pictures in the ad.  
- `postalCode` - The postal code for the location of the vehicle.  
- `lastSeenOnline` - When the crawler saw this ad last online.  

In [97]:
import pandas as pd
import numpy as np

Read in the dataset:

In [98]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset contains 50,000 rows and 20 columns.  Most of the data is stored as strings.  Some columns have null values, but no columns have more than 20% null values.  

In [99]:
# check to see the % of null values in each column.
autos.apply(pd.isnull).sum()/autos.shape[0]

dateCrawled            0.00000
name                   0.00000
seller                 0.00000
offerType              0.00000
price                  0.00000
abtest                 0.00000
vehicleType            0.10190
yearOfRegistration     0.00000
gearbox                0.05360
powerPS                0.00000
model                  0.05516
odometer               0.00000
monthOfRegistration    0.00000
fuelType               0.08964
brand                  0.00000
notRepairedDamage      0.19658
dateCreated            0.00000
nrOfPictures           0.00000
postalCode             0.00000
lastSeen               0.00000
dtype: float64

### 3. Clean the Columns

Clean the column names to make the data easier to work with.  
The columns names are in camelcase instead of snakecase.
* Convert the column names from camelcase to snakecase and adjust some column names based on the data dictionary to add more description.

In [100]:
# print an array of the existing column names
new_col_names = []
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [101]:
# function to convert from camelcase to snakecase
def camel_to_snake(str):
    return ''.join(['_'+i.lower() if i.isupper()
                    else i for i in str]).lstrip('_')

In [102]:
# change column names to give a more accurate description and convert to snakecase
for name in autos.columns:
    name = name.replace("yearOfRegistration","registration_year")
    name = name.replace("monthOfRegistration","registration_month")
    name = name.replace("notRepairedDamage","unrepaired_damage")
    name = name.replace("dateCreated","ad_created")
    name = name.replace("powerPS","power_ps")
    name = camel_to_snake(name)
    new_col_names.append(name)

In [103]:
new_col_names

['date_crawled',
 'name',
 'seller',
 'offer_type',
 'price',
 'abtest',
 'vehicle_type',
 'registration_year',
 'gearbox',
 'power_ps',
 'model',
 'odometer',
 'registration_month',
 'fuel_type',
 'brand',
 'unrepaired_damage',
 'ad_created',
 'nr_of_pictures',
 'postal_code',
 'last_seen']

In [104]:
# assign the modified column names back to the autos.columns attribute
autos.columns = new_col_names

In [105]:
# check the new column names 
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

## 4. Initial Data Exploration and Cleaning
* Check for columns where all or most of the values are the same.
* Check for numeric data stored as text which will need to be cleaned and converted.


In [106]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

**Columns `"seller"` & `"offer_type"` contain the same value and can be removed**

In [107]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [108]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [109]:
autos.drop(["seller","offer_type"], axis = 1, inplace = True)

**The `"nr_of_pictures"` column has 0 for every entry.  Also drop this column.**

In [110]:
autos.drop(["nr_of_pictures"], axis = 1, inplace= True)

**Columns `"price"` & `"odometer"` are stored as text and contain non-numeric characters.  This should be removed and values converted to a numeric dtype.**

In [111]:
autos["price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [112]:
autos["odometer"].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

In [113]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

/tmp/ipykernel_893008/2855709183.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)


In [114]:
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [115]:
autos["odometer"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

**Rename the `"odometer"` column to `"odometer_km"`**

In [116]:
autos.rename(columns = {"odometer":"odometer_km"}, inplace=True)

In [117]:
# check to confirm that the column was renamed
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

## 5. Further Explore the Odometer and Price Columns
* Check for data that does not look correct and search for any outliers.
* Remove outliers and provide evidence why removed.
    - Check for min and max values and check if any values look too low or too high


In [118]:
price = autos["price"]
adometer_km = autos["odometer_km"]

In [119]:
price.unique().shape

(2357,)

In [120]:
price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [121]:
price.value_counts().sort_index(ascending = False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

In [122]:
price.value_counts().sort_index(ascending = True).head(100)

0      1421
1       156
2         3
3         1
5         2
       ... 
300     384
310       1
320      12
325       5
329       2
Name: price, Length: 100, dtype: int64

In [123]:
autos.loc[autos["price"] == 12345678]

date_crawled                                               name  \
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   

          price   abtest vehicle_type  registration_year  gearbox  power_ps  \
27371  12345678  control          NaN               2017      NaN        95   
39377  12345678  control          NaN               2018  manuell        95   
47598  12345678  control    limousine               2001  manuell       101   

        model  odometer_km  registration_month fuel_type  brand  \
27371   punto       150000                   0       NaN   fiat   
39377     v40       150000                   6       NaN  volvo   
47598  vectra       150000                   3    benzin   opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
27371               NaN  2016-03-09 00:00:00        96110  2016-03-09 15:45:47  
39377              nein  2016-03-08 00:00:00        14542  2016-04-06 23:17:31  
47598              nein  2016-03-31 00:00:00         4356  2016-03-31 18:56:54

**There are 1,412 cars listed with a price of $0.  This represents around ~2% and can be considered to be removed.  The max price is $100,000,000 and can also be removed.**  

**Remove outliers in the data with prices between 200 to 400,000**

In [124]:
autos = autos[autos["price"].between(200,400000)]

In [125]:
autos["price"].describe()

count     47645.000000
mean       6000.707273
std        9110.783444
min         200.000000
25%        1300.000000
50%        3190.000000
75%        7500.000000
max      350000.000000
Name: price, dtype: float64

## 6. Exploring the Date Columns

**There are seveal columns that represent date information;**  
1. 'date_crawled'
2. 'last_seen'
3. 'ad_created'
4. 'registration_month'
5. 'registration_year'

**3 of the 5 fields are formated as strings.**
Check their contents and extract just the date
1. 'date_crawled'
2. 'last_seen'
3. 'ad_created'

In [126]:
autos[['date_crawled','last_seen','ad_created','registration_month','registration_year']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 47645 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        47645 non-null  object
 1   last_seen           47645 non-null  object
 2   ad_created          47645 non-null  object
 3   registration_month  47645 non-null  int64 
 4   registration_year   47645 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 2.2+ MB


In [127]:
# show the first six rows of the date columns
autos[['date_crawled','last_seen','ad_created']][0:6]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00
5  2016-03-21 13:47:45  2016-04-06 09:45:21  2016-03-21 00:00:00

In [128]:
autos["date_crawled"].value_counts().sort_index(ascending = False)

2016-04-07 14:36:56    1
2016-04-07 14:36:55    1
2016-04-07 14:36:44    1
2016-04-07 14:30:26    1
2016-04-07 14:30:09    1
                      ..
2016-03-05 14:07:26    1
2016-03-05 14:07:21    1
2016-03-05 14:07:08    1
2016-03-05 14:06:40    1
2016-03-05 14:06:30    1
Name: date_crawled, Length: 46029, dtype: int64

In [129]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = False)

2016-04-07    0.001364
2016-04-06    0.003169
2016-04-05    0.013139
2016-04-04    0.036562
2016-04-03    0.038724
2016-04-02    0.035681
2016-04-01    0.033813
2016-03-31    0.031840
2016-03-30    0.033897
2016-03-29    0.033980
2016-03-28    0.035030
2016-03-27    0.031210
2016-03-26    0.032217
2016-03-25    0.031420
2016-03-24    0.029321
2016-03-23    0.032385
2016-03-22    0.032700
2016-03-21    0.037360
2016-03-20    0.037800
2016-03-19    0.034610
2016-03-18    0.012824
2016-03-17    0.031546
2016-03-16    0.029447
2016-03-15    0.034232
2016-03-14    0.036562
2016-03-13    0.015699
2016-03-12    0.036877
2016-03-11    0.032700
2016-03-10    0.032322
2016-03-09    0.033036
2016-03-08    0.033120
2016-03-07    0.035995
2016-03-06    0.014062
2016-03-05    0.025354
Name: date_crawled, dtype: float64

**It looks like the data was compiled overa a one month period from March to April 2016.  The distribution of the listings is roughly uniform for each day.**

In [130]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = False)

2016-04-07    0.001217
2016-04-06    0.003253
2016-04-05    0.011859
2016-04-04    0.036919
2016-04-03    0.038976
                ...   
2015-12-05    0.000021
2015-11-10    0.000021
2015-09-09    0.000021
2015-08-10    0.000021
2015-06-11    0.000021
Name: ad_created, Length: 76, dtype: float64

In [131]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = False)

2016-04-07    0.132669
2016-04-06    0.222437
2016-04-05    0.125470
2016-04-04    0.024620
2016-04-03    0.025102
2016-04-02    0.024808
2016-04-01    0.022899
2016-03-31    0.023864
2016-03-30    0.024578
2016-03-29    0.022185
2016-03-28    0.020758
2016-03-27    0.015511
2016-03-26    0.016686
2016-03-25    0.019079
2016-03-24    0.019624
2016-03-23    0.018491
2016-03-22    0.021408
2016-03-21    0.020506
2016-03-20    0.020653
2016-03-19    0.015615
2016-03-18    0.007283
2016-03-17    0.028083
2016-03-16    0.016287
2016-03-15    0.015720
2016-03-14    0.012530
2016-03-13    0.008899
2016-03-12    0.023906
2016-03-11    0.012299
2016-03-10    0.010473
2016-03-09    0.009613
2016-03-08    0.007178
2016-03-07    0.005373
2016-03-06    0.004303
2016-03-05    0.001091
Name: last_seen, dtype: float64

In [132]:
autos["registration_year"].value_counts().sort_index().head()

1000    1
1001    1
1111    1
1800    2
1910    2
Name: registration_year, dtype: int64

## 7. Dealing with Incorrect Registration Year Data

**Check the *registration_year* column and clean the data**
* The min value is 1000 - remove
* The max value is 9999 and some entries are after the listing year (2016)

In [133]:
autos["registration_year"].value_counts().sort_index(ascending = False).head(20)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       3
4800       1
4500       1
4100       1
2800       1
2019       1
2018     468
2017    1371
2016    1161
2015     376
2014     656
2013     797
2012    1307
2011    1617
2010    1587
Name: registration_year, dtype: int64

In [134]:
autos["registration_year"].value_counts().sort_index(ascending = True).head(20)

1000    1
1001    1
1111    1
1800    2
1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    1
1951    2
1952    1
1953    1
1954    2
Name: registration_year, dtype: int64

In [135]:
# check some of the details corresponding to very old dates for the registration_year
autos.loc[autos["registration_year"] == 1927]

date_crawled                     name  price   abtest  \
21416  2016-03-12 08:36:21  Essex_super_six__Ford_A  16500  control   

      vehicle_type  registration_year  gearbox  power_ps   model  odometer_km  \
21416       cabrio               1927  manuell        40  andere         5000   

       registration_month fuel_type brand unrepaired_damage  \
21416                   5    benzin  ford               NaN   

                ad_created  postal_code            last_seen  
21416  2016-03-12 00:00:00        74821  2016-03-15 12:45:12

In [136]:
# remove entries in the dataframe with registration_year between 1926 to 2016
autos = autos[autos["registration_year"].between(1926,2016)]

**Most of the cars listed were made in the 2000's**

In [137]:
# caculate the distribution of the remaining values in the DF
autos["registration_year"].value_counts(normalize=True).head(20)

2000    0.065940
2005    0.063472
1999    0.062140
2004    0.058842
2003    0.058798
2006    0.058252
2001    0.057029
2002    0.053949
1998    0.049908
2007    0.049624
2008    0.048183
2009    0.045409
1997    0.040451
2011    0.035318
2010    0.034663
2012    0.028547
1996    0.028438
2016    0.025358
1995    0.024441
2013    0.017408
Name: registration_year, dtype: float64

## 8. Exploring Price by Brand


**Use aggeration of the top 20 brands to understand details of the *'brand'* column**


In [138]:
top_20 = autos["brand"].value_counts()[:20]
top_20

volkswagen        9672
bmw               5094
opel              4848
mercedes_benz     4470
audi              4011
ford              3147
renault           2130
peugeot           1371
fiat              1149
seat               832
skoda              756
nissan             700
mazda              693
smart              658
citroen            647
toyota             592
hyundai            463
sonstige_autos     437
volvo              421
mini               407
Name: brand, dtype: int64

**Loop over the top 20 brands and assign the mean price to the dictionary *top_brands***

In [139]:
top_brands = {}
top_brands_keys = top_20.index
top_brands_keys
#top_brands = top_brands[top_brand_keys]


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [146]:
for maker in top_brands_keys:
    avg_price = autos.loc[autos["brand"] == maker,"price"].mean()
    top_brands[maker] = int(avg_price)

print(top_brands)

{'volkswagen': 5506, 'bmw': 8402, 'opel': 3078, 'mercedes_benz': 8691, 'audi': 9406, 'ford': 3883, 'renault': 2553, 'peugeot': 3142, 'fiat': 2925, 'seat': 4505, 'skoda': 6451, 'nissan': 4829, 'mazda': 4204, 'smart': 3596, 'citroen': 3818, 'toyota': 5175, 'hyundai': 5422, 'sonstige_autos': 12929, 'volvo': 5016, 'mini': 10665}


### Sorting Aggregate Data in a DataFrame

In [147]:
bmp_series = pd.Series(top_brands)

In [148]:
bmp_series.sort_values(ascending=False)

sonstige_autos    12929
mini              10665
audi               9406
mercedes_benz      8691
bmw                8402
skoda              6451
volkswagen         5506
hyundai            5422
toyota             5175
volvo              5016
nissan             4829
seat               4505
mazda              4204
ford               3883
citroen            3818
smart              3596
peugeot            3142
opel               3078
fiat               2925
renault            2553
dtype: int64

## 9. Conclusion
Summary - There are many listing on the site and in cleaning the data, several rows and columns were removed.  Once the dataset was cleaned it was possible to gain insites from looking at the data.

- Most of the German vehicles belonging to large brands tended to cost more than common used counterparts.